jaiqu/jaiqu_burr_granular.py

```diff
@@ -0,0 +1,237 @@
+                state = state.update(max_retries_hit=True, complete_filter=complete_filter)
+                return {
+                    "error": f"Failed to validate the jq filter after {max_retries} retries."}, state
+
+            complete_filter = repair_query(complete_filter, str(e), input_json, None)
+    state = state.update(complete_filter=complete_filter, max_retries_hit=False)
+    return {"complete_filter": complete_filter}, state
+
+
+def translate_schema(input_json, output_schema, openai_api_key: str | None = None, key_hints=None,
+                     max_retries=10) -> str:
+    app = build_application(input_json, output_schema, openai_api_key, key_hints, max_retries)
+    last_action, result, state = app.run(halt_after=["error_state", "good_result"])
+    if last_action == "error_state":
+        raise RuntimeError(result)
+    return result["complete_filter"]
+
+
+def build_application(input_json="", output_schema="", openai_api_key: str | None = None, key_hints=None,
+                      max_retries=10):
+    app = (
+        ApplicationBuilder()
+        .with_state(
+            **{
+                "input_json": input_json,
+                "output_schema": output_schema,
+                "key_hints": key_hints,
+                "max_retries": max_retries,
+            }
+        )
+        .with_actions(
+            # bind the vector store to the AI conversational step
+            validate_schema=validate_schema,
+            create_jq_filter_query=create_jq_filter_query,
+            validate_json=validate_json,
+            finalize_filter=finalize_filter,
+            error_state=burr.core.Result("complete_filter"),
+            good_result=burr.core.Result("complete_filter"),
+        )
+        .with_transitions(
+            ("validate_schema", "validate_schema", when(schema_processed=False)),
+            ("validate_schema", "create_jq_filter_query", when(schema_processed=True)),
+            ("validate_schema", "error_state", when(valid_schema=False)),
+            ("create_jq_filter_query", "create_jq_filter_query", when(filters_created=False)),
+            ("create_jq_filter_query", "finalize_filter", when(filters_created=True)),
+            ("create_jq_filter_query", "error_state", when(max_retries_hit=True)),
+            ("finalize_filter", "validate_json", default),
+            ("validate_json", "error_state", when(max_retries_hit=True)),
+            ("validate_json", "good_result", default),
+        )
+        .with_entrypoint("validate_schema")
+        .with_tracker(project="example:jaiqu")
+        .with_identifiers(partition_key="dagworks")
+        .build()
+    )
+    app.visualize(
+        output_file_path="jaiqu_granular", include_conditions=True, view=True, format="png"
+    )
+    return app
+
+
+if __name__ == '__main__':
+    # app = build_application()
+    # app.visualize(
+    #     output_file_path="jaiqu_granular", include_conditions=True, view=True, format="png"
+    # )
```

Comment:
Stated above, but main code with dummy data is probably better served as an example file.